app/schema.py

```diff
@@ -55,9 +80,95 @@ def mutate(root, info, contact, password):
+        except User.DoesNotExist:
+            return SendFriendRequest(message="user not found", ok=False, user_receiver=None, friend_request=None)
+
+        try:
+            friend_request = FriendRequest.objects.get(from_user=info.context.user, to_user=user_receiver)
+            return SendFriendRequest(message="A friend request has already been sent", ok=False, user_receiver=user_receiver, friend_request=None)
+
+        except FriendRequest.DoesNotExist:
+            if FriendRequest.objects.filter(from_user=user_receiver, to_user=info.context.user).exists():
+                return SendFriendRequest(message="User has sent a request before", ok=False, user_receiver=user_receiver, friend_request=None)
+
+            if Friend.objects.filter(current_user=info.context.user, friend=user_receiver).exists():
+                return SendFriendRequest(message="Both users are already friends", ok=False, user_receiver=user_receiver, friend_request=None)
+            
+            friend_request = FriendRequest.objects.create(from_user=info.context.user, to_user=user_receiver)
+            return SendFriendRequest(message="Friend request sent successfully", ok=True, user_receiver=user_receiver, friend_request=friend_request)
+
+
+class AcceptFriendRequest(graphene.Mutation):
+    class Arguments:
+        contact_sender = graphene.String()
+    ok = graphene.Boolean()
+    user_sender = graphene.Field(UserType)
+    friend_request = graphene.Field(FriendRequestType)
+    friend = graphene.Field(FriendType)
+
+    @staticmethod
+    @login_required
+    def mutate(root, info, contact_sender):
+        try:
+            user_sender = User.objects.get_by_natural_key(contact_sender)
+
+        except User.DoesNotExist:
+            return AcceptFriendRequest(ok=False, user_sender=None, friend_request=None, friend=None)
+
+        try:
+            friend_request = FriendRequest.objects.get(from_user=user_sender, to_user=info.context.user)
+            friend = Friend.accept(current_user=info.context.user, friend=user_sender)
+            return AcceptFriendRequest(ok=True, user_sender=user_sender, friend_request=friend_request, friend=friend)
+
+        except FriendRequest.DoesNotExist:
+            return AcceptFriendRequest(ok=False, user_sender=user_sender, friend_request=None, friend=None)
+
+
+class DeleteFriendRequest(graphene.Mutation):
+    class Arguments:
+        contact_sender = graphene.String()
+    ok = graphene.Boolean()
+    user_sender = graphene.Field(UserType)
+    friend_request = graphene.Field(FriendRequestType)
+
+    @staticmethod
+    @login_required
+    def mutate(root, info, contact_sender):
+        try:
+            user_sender = User.objects.get_by_natural_key(contact_sender)
+
+        except User.DoesNotExist:
+            return DeleteFriendRequest(ok=False, user_sender=None, friend_request=None)
+
+        try:
+            friend_request = FriendRequest.objects.get(from_user=user_sender, to_user=info.context.user)
```

Comment:
The friend_request can be sent by the user also. For example, earlier I sent a friend request, but now I want to delete it. You are only handling the case where the user can delete requests which he has received

Reply:
To delete received friend requests only the contact of the sender was required as the contact of the receiver was assumed to be the request user's.

To incorporate the deletion of the received friend requests in the same mutation, I will have to take two inputs (contact of sender and contact of receiver)

Is that okay?

Reply:
I guess I wasn't clear about this part, but that's okay